includes/wikia/api/TemplateClassificationApiController.class.php

```diff
@@ -0,0 +1,177 @@
+			$type = $this->getApiClient()->getTemplateType( $wikiId, $pageId );
+			if ( !is_null( $type ) ) {
+				$templateType = $type->getType();
+			}
+		} catch ( InvalidArgumentException $e ) {
+			throw new BadRequestApiException( $e->getMessage() );
+		} catch ( ApiException $e ) {
+			throw $e;
+		}
+
+		return $templateType;
+	}
+
+	/**
+	 * Get details about template type
+	 *
+	 * @param int $wikiId
+	 * @param int $pageId
+	 * @return array
+	 * @throws ApiException
+	 * @throws BadRequestApiException
+	 * @throws Exception
+	 */
+	public function getDetails( $wikiId, $pageId ) {
+		$templateDetails = [];
+
+		try {
+			$details = $this->getApiClient()->getTemplateDetails( $wikiId, $pageId );
+
+			if ( !is_null($details) ) {
+				$providers = $details->getProviders();
+				$templateDetails = $this->prepareTemplateDetails( $providers );
+			}
+		} catch ( InvalidArgumentException $e ) {
+			throw new BadRequestApiException( $e->getMessage() );
+		} catch ( ApiException $e ) {
+			throw $e;
+		}
+
+		return $templateDetails;
+	}
+
+	/**
+	 * Classify template type
+	 *
+	 * @param int $userId
+	 * @param int $wikiId
+	 * @param int $pageId
+	 * @param string $templateType
+	 * @throws ApiException
+	 * @throws BadRequestApiException
+	 * @throws Exception
+	 * @throws PermissionsException
+	 * @throws UnauthorizedException
+	 */
+	public function classifyType( $wikiId, $pageId, $templateType ) {
+		$this->validateRequest( $this->wg->User, $this->request );
+
+		$details = [
+			'provider' => self::USER_PROVIDER,
+			'origin' => $this->wg->User->getId(),
+			'types' => [ $templateType ]
+		];
+		$templateTypeProvider = new TemplateTypeProvider( $details );
+
+		try {
+			$this->getApiClient()->insertTemplateDetails( $wikiId, $pageId, $templateTypeProvider );
+		} catch ( InvalidArgumentException $e ) {
+			throw new BadRequestApiException( $e->getMessage() );
+		} catch ( ApiException $e ) {
+			throw $e;
+		}
+	}
+
+	/**
+	 * Prepare template details output
+	 *
+	 * @param TemplateTypeProvider[] $providers
+	 * @return array
+	 */
+	private function prepareTemplateDetails( $providers ) {
+		$templateDetails = [];
+
+		foreach ( $providers as $provider ) {
+			$templateDetails[$provider->getProvider()] = [
+				'provider' => $provider->getProvider(),
+				'origin' => $provider->getOrigin(),
+				'types' => $provider->getTypes()
+			];
+		}
+
+		return $templateDetails;
+	}
+
+	/**
+	 * Check permission to classify template type
+	 *
+	 * @param User $user
+	 * @param WikiaRequest $request
+	 * @return bool
+	 * @throws PermissionsException
+	 * @throws UnauthorizedException
+	 * @throws BadRequestException
+	 */
+	private function validateRequest( User $user, WikiaRequest $request ) {
+		if ( !$request->wasPosted() ) {
+			throw new BadRequestApiException();
+		}
+
+		if ( !$user->isLoggedIn() || !$user->matchEditToken( $request->getVal( 'editToken' ) ) ) {
+			throw new UnauthorizedException();
+		}
+
+		if ( !$user->isAllowed( 'edit' ) ) {
```

Comment:
Should we be checking if they can edit the page they are classifying? Which would also check if the user is blocked.